lib/rucio/core/transfer.py

```diff
@@ -721,381 +1058,239 @@ def get_transfer_requests_and_source_replicas(total_workers=0, worker_number=0,
+                reqs_scheme_mismatch.add(rws.request_id)
+                reqs_no_source.remove(rws.request_id)
+
+    # Ensure each single-hop transfer has its own request id. Create intermediate replicas and requests if needed.
+    # Organize into the return format: {request_id: <transfer_definition_dict>}
+    transfers = {}
+    for rws, transfer_path in transfer_path_for_request:
+        # The last hop is the initial request
+        transfers[rws.request_id] = transfer_path[-1]
+        transfers[rws.request_id]['file_metadata']['request_id'] = rws.request_id
+        transfers[rws.request_id]['request_id'] = rws.request_id
 
-                if current_source_is_tape and new_source_is_tape:
-                    # multiple Tape source replicas are not allowed in FTS3.
-                    # Either keep the old source. Or substitute it with the new one.
-                    prev_is_multihop = transfers[req_id]['sources'][0][4] is None  # will be None if the previous transfer is multihop
-                    if ranking > transfers[req_id]['sources'][0][3]\
-                            or (ranking == transfers[req_id]['sources'][0][3] and (prev_is_multihop
-                                                                                   or link_ranking < transfers[req_id]['sources'][0][4])):
-                        transfers[req_id]['sources'] = []
-                        transfers[req_id]['bring_online'] = bring_online_local
-                        transfers[req_id]['file_metadata']['src_rse'] = rse
-                    else:
-                        continue
+        # If the transfer is a multihop, need to create the intermediate replicas, intermediate requests and the transfers
+        parent_request = None
+        parent_requests = []
+        for transfer in transfer_path[:-1]:
+            # hop = {'source_rse_id': source_rse_id, 'source_scheme': 'srm', 'source_scheme_priority': N, 'dest_rse_id': dest_rse_id, 'dest_scheme': 'srm', 'dest_scheme_priority': N}
+            source_rse_id = transfer['file_metadata']['src_rse_id']
+            dest_rse_id = transfer['file_metadata']['dest_rse_id']
+            source_rse_name = ctx.rse_name(source_rse_id)
+            dest_rse_name = ctx.rse_name(dest_rse_id)
+            dest_rse_vo = get_rse_vo(rse_id=dest_rse_id, session=session)
+
+            files = [{'scope': rws.scope,
+                      'name': rws.name,
+                      'bytes': rws.byte_count,
+                      'adler32': rws.adler32,
+                      'md5': rws.md5,
+                      'state': 'C'}]
+            try:
+                add_replicas(rse_id=dest_rse_id,
+                             files=files,
+                             account=InternalAccount('root', vo=dest_rse_vo),
+                             ignore_availability=False,
+                             dataset_meta=None,
+                             session=session)
+            except Exception as error:
```

Comment:
In case of exception, the session is rollback if I'm not wrong

Reply:
@bari12 , any opinion ? It's not a behavior which changed. It looks like new code because github detects the diff wrongly. But this behavior is old (doesn't mean it's correct though)

Reply:
This is fine. It is the same as before. It actually checks if there already is a replica in `add_replicas`. For the case that it did not detect it at that point, but due to another session adding one exactly at this time, this will fire an integrity-constraint error, and the session will be rolled back. But also that should be fine.